mm2src/gui_storage/src/rpc_commands.rs

```diff
@@ -0,0 +1,381 @@
+    #[serde(flatten)]
+    account: NewAccount<AccountId>,
+}
+
+#[derive(Deserialize)]
+pub struct DeleteAccountRequest {
+    account_id: AccountId,
+}
+
+#[derive(Deserialize)]
+pub struct SetAccountNameRequest {
+    account_id: AccountId,
+    name: String,
+}
+
+#[derive(Deserialize)]
+pub struct SetAccountDescriptionRequest {
+    account_id: AccountId,
+    description: String,
+}
+
+#[derive(Deserialize)]
+pub struct CoinRequest {
+    account_id: AccountId,
+    tickers: Vec<String>,
+}
+
+#[derive(Deserialize)]
+pub struct GetAccountsRequest;
+
+#[derive(Deserialize)]
+pub struct GetAccountCoinsRequest {
+    account_id: AccountId,
+}
+
+#[derive(Serialize)]
+pub struct GetAccountCoinsResponse {
+    account_id: AccountId,
+    coins: BTreeSet<String>,
+}
+
+#[derive(Deserialize)]
+pub struct GetEnabledAccountRequest;
+
+#[derive(Deserialize)]
+pub struct SetBalanceRequest {
+    account_id: AccountId,
+    balance_usd: BigDecimal,
+}
+
+/// Sets the given account as an enabled (current active account).
+/// The behaviour depends on [`EnableAccountRequest::policy`]:
+/// * [`EnableAccountPolicy::Known`] =>
+///     1) Checks whether the given account exists in the storage.
+///        Returns [`AccountRpcError::NoSuchAccount`] if there is no account with the given `AccountId`;
+///     2) Sets the account as an enabled.
+/// * [`EnableAccountPolicy::New`] =>
+///     1) Tries to upload the given account info to the storage.
+///        Returns [`AccountRpcError::AccountExistsAlready`] if there is an account with the same `AccountId` already;
+///     2) Sets the account as an enabled.
+///
+/// # Important
+///
+/// This RPC affects the storage **only**. It doesn't affect MarketMaker.
+pub async fn enable_account(ctx: MmArc, req: EnableAccountRequest) -> MmResult<SuccessResponse, AccountRpcError> {
+    let account_ctx = AccountContext::from_ctx(&ctx).map_to_mm(AccountRpcError::Internal)?;
+    let account_id = match req.policy {
+        EnableAccountPolicy::Existing(account_id) => account_id,
+        EnableAccountPolicy::New(new_account) => {
+            let account_id = new_account.account_id;
+            account_ctx
+                .storage()
+                .await?
+                .upload_account(AccountInfo::from(new_account))
+                .await?;
+            account_id
+        },
+    };
+    account_ctx.storage().await?.enable_account(account_id).await?;
+    Ok(SuccessResponse::new())
+}
+
+/// Adds the given [`AddAccountRequest::account`] to the storage.
+/// Returns [`AccountRpcError::AccountExistsAlready`] if there is an account with the same `AccountId` already.
+///
+/// # Important
+///
+/// This RPC affects the storage **only**. It doesn't affect MarketMaker.
+pub async fn add_account(ctx: MmArc, req: AddAccountRequest) -> MmResult<SuccessResponse, AccountRpcError> {
+    let account_ctx = AccountContext::from_ctx(&ctx).map_to_mm(AccountRpcError::Internal)?;
+    validate_new_account(&req.account)?;
```

Comment:
Validation should be above of `let account_ctx` There is no point getting `AccountContext` if the payload isn't valid.

Reply:
Makes sense, will do

Reply:
Done